evap/staff/views.py

```diff
@@ -801,12 +816,52 @@ def questionnaire_copy(request, questionnaire_id):
         else:
             return render(request, "staff_questionnaire_form.html", dict(form=form, formset=formset))
     else:
-        questionnaire = get_object_or_404(Questionnaire, id=questionnaire_id)
-        InlineQuestionFormset = inlineformset_factory(Questionnaire, Question, formset=AtLeastOneFormSet, form=QuestionForm, extra=1, exclude=('questionnaire',))
+        form, formset = get_identical_form_and_formset(copied_questionnaire)
+        return render(request, "staff_questionnaire_form.html", dict(form=form, formset=formset))
 
-        form = QuestionnaireForm(instance=questionnaire)
-        formset = InlineQuestionFormset(instance=questionnaire, queryset=questionnaire.question_set.all())
 
+@staff_required
+def questionnaire_new_version(request, questionnaire_id):
+    old_questionnaire = get_object_or_404(Questionnaire, id=questionnaire_id)
+
+    # Check if we can use the old name with the current time stamp.
```

Comment:
i'd add a `if old_questionnaire.is_obsolete: raise PermissionDenied`. this "protects" against users manually forging URLs or pressing the back button after creating a new version, but is actually more for documenting that this is not intended.